### DISCUSSION TITLE
Android TV DPad longPress event

### DISCUSSION THREAD

**Discussion by fulminant** [react-native-tvos/react-native-tvos#682] [General Discussions]
Greetings,

I hope this message finds you well.

Firstly, I extend my heartfelt appreciation for the multitude of features you've seamlessly integrated into the TV platform. Your dedication to enhancing user experience does not go unnoticed.

During my exploration of the codebase, I stumbled upon a captivating discovery, one that appears to be deliberately embedded within the code's intricate fabric.

In my pursuit to replicate the functionality of a contextual menu (modal) invocation akin to that of the YouTube TV application, I endeavored to trigger its appearance through a prolonged press on the directional pad (DPad).

While delving into the React Native (RN) documentation, it became evident that there exists a semblance of support for the onLongPress event. However, despite numerous attempts employing various touchable components such as Touchable, TouchableOpacity, TouchableHigh, and the like, regrettably, the desired functionality remained elusive.

Upon thorough examination, I discerned that the introduction of TV support was heralded by a significant commit, as evidenced by the following link: [GitHub Commit](https://github.com/react-native-tvos/react-native-tvos/commit/98f21377a2252ebd356529d1f9af8e751a6b5cf7).

It appears that all Touchable components, inclusive of their implementation of the onLongPress event, are characterized by a conditional check for the Android platform:

```
onPress: event => {
    if (this.props.onPress != null && Platform.OS !== 'android') {
        this.props.onPress(event);
    }
},
onLongPress: event => {
    if (this.props.onLongPress != null && Platform.OS !== 'android') {
        this.props.onLongPress(event);
    }
},
```

Consequently, due to this validation specifically tailored for Android devices, the propagation of the onLongPress event is impeded.

While my proficiency in Kotlin and native development may be somewhat lacking, I am inclined to believe that this design choice was deliberate in nature. However, the rationale behind such a decision eludes my comprehension.

Thus, I humbly seek your guidance on the feasibility of triggering onLongPress events on Android TV. Any insights or assistance you could provide would be immensely appreciated.

Thank you for your time and consideration.

**Comment by douglowder** [comment]
Well it wasn't really a design choice, it was because on Android, those events are fired elsewhere, and we didn't want duplicate events being fired.

I think you've raised a valid point... this probably needs to be looked at and the behavior made more consistent between the two platforms. I'll convert this to an issue.

All that said, I do recommend that if you are starting out building a new app, you probably want to use the more modern `Pressable` component, and not the `Touchable` components.

**Comment by fulminant** [comment]
Thank you for getting back to me.
I'm pleased that my question made sense.

I've been exploring the `Pressable` component in documentation, but I'm having trouble getting it to function correctly (focus, press) with the remote (D-Pad) on Android TV. Do you have any tips or advice on this?

Thanks in advance.

**Comment by fulminant** [comment]
Sorry, my bad. Focus event works correctly.

But the `longPress` event still doesn't work.
After some debugging, I found this piece of code [packages/react-native/Libraries/Components/Pressable/Pressable.js](https://github.com/react-native-tvos/react-native-tvos/blob/a4922d19b8c9879be10b671fab81f99e2ed3fa8f/packages/react-native/Libraries/Components/Pressable/Pressable.js#L393) which conditionally checks is it `tvOS` and doesn't emit the `longPress` event on the AndroidTV.

**Comment by douglowder** [comment]
@fulminant I had some time to look into this, and figured out the actual issue that was producing duplicate calls of the event handlers in `Touchable` components.

Fix is in https://github.com/react-native-tvos/react-native-tvos/pull/690

**Reply by fulminant** [reply]
@douglowder, Thank you so much for taking the time to investigate and resolve the issue with the duplicate calls of the event handlers in the Touchable components.
